service/worker/sysworkflow/system_workflow.go

```diff
@@ -22,101 +22,341 @@ package sysworkflow
-	logger := workflow.GetLogger(ctx)
-	scope := workflow.GetMetricsScope(ctx).Tagged(map[string]string{SystemWorkflowIDTag: id})
-	ch := workflow.GetSignalChannel(ctx, SignalName)
-
-	logger.Info("started new system workflow")
+	sysWorkflowInfo := workflow.GetInfo(ctx)
+	logger := globalLogger.WithFields(bark.Fields{
+		logging.TagWorkflowExecutionID: sysWorkflowInfo.WorkflowExecution.ID,
+		logging.TagWorkflowRunID:       sysWorkflowInfo.WorkflowExecution.RunID,
+	})
+	logger.Info("started system workflow")
+	globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerWorkflowStarted)
+	sw := globalMetricsClient.StartTimer(metrics.SystemWorkflowScope, metrics.SysWorkerContinueAsNewLatency)
+	ch := workflow.GetSignalChannel(ctx, signalName)
 	signalsHandled := 0
-	for ; signalsHandled < SignalsUntilContinueAsNew; signalsHandled++ {
+	for ; signalsHandled < signalsUntilContinueAsNew; signalsHandled++ {
 		var signal signal
 		if more := ch.Receive(ctx, &signal); !more {
-			scope.Counter(ChannelClosedUnexpectedlyError).Inc(1)
-			logger.Error("cadence channel was unexpectedly closed")
 			break
 		}
-		selectSystemTask(scope, signal, ctx, logger)
+		globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerReceivedSignal)
+		selectSystemTask(signal, ctx, logger, globalMetricsClient)
 	}
 
 	for {
 		var signal signal
 		if ok := ch.ReceiveAsync(&signal); !ok {
 			break
 		}
-		selectSystemTask(scope, signal, ctx, logger)
+		globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerReceivedSignal)
+		selectSystemTask(signal, ctx, logger, globalMetricsClient)
 		signalsHandled++
 	}
-
-	logger.Info("completed current set of iterations, continuing as new",
-		zap.Int(logging.TagIterationsUntilContinueAsNew, signalsHandled))
-
-	ctx = workflow.WithExecutionStartToCloseTimeout(ctx, WorkflowStartToCloseTimeout)
-	ctx = workflow.WithWorkflowTaskStartToCloseTimeout(ctx, DecisionTaskStartToCloseTimeout)
-	return workflow.NewContinueAsNewError(ctx, SystemWorkflowFnName)
+	ctx = workflow.WithExecutionStartToCloseTimeout(ctx, workflowStartToCloseTimeout)
+	ctx = workflow.WithWorkflowTaskStartToCloseTimeout(ctx, decisionTaskStartToCloseTimeout)
+	globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerContinueAsNew)
+	sw.Stop()
+	logger.WithFields(bark.Fields{
+		logging.TagNumberOfSignalsUntilContinueAsNew: signalsHandled,
+	}).Info("system workflow is continuing as new")
+	return workflow.NewContinueAsNewError(ctx, systemWorkflowFnName)
 }
 
-func selectSystemTask(scope tally.Scope, signal signal, ctx workflow.Context, logger *zap.Logger) {
-	scope.Counter(HandledSignalCount).Inc(1)
-
+func selectSystemTask(signal signal, ctx workflow.Context, logger bark.Logger, metricsClient metrics.Client) {
 	ao := workflow.ActivityOptions{
 		ScheduleToStartTimeout: time.Minute,
 		StartToCloseTimeout:    time.Minute,
 		HeartbeatTimeout:       time.Second * 10,
```

Comment:
I am not sure I understand. If you do not heartbeat for 10 seconds the activity will time out and be retried. When you try to upload a blob you will always retry forever as long as you are getting transient errors. So if you get transient errors for too long then this heartbeat will kick in and kill your activity, at which point the activity will be retried from the start again.

Reply:
Basically the way I thought about it is, the activity retries everything it can forever, and heartbeat acts as a catch all to say "if you have not made enough forward progress in last 10 seconds, then start all the way over"